### ISSUE TITLE
Support for Cockroach DB

### ISSUE THREAD

**Issue by kylepl** [cyrilgdn/terraform-provider-postgresql#372]
Hi,

I'm attempting to use this provider with Cockroach DB, which aims to be [~Postgres compatible](https://www.cockroachlabs.com/docs/stable/postgresql-compatibility). A few different bugs show that features relied upon by cockroach DB:
* https://github.com/cockroachdb/cockroach/issues/39109
* https://github.com/cyrilgdn/terraform-provider-postgresql/issues/277

I've managed to hack-up a version such that I can at least support the `Role` resource, and tested it locally ([commit](https://github.com/kylepl/terraform-provider-postgresql/commit/9c2eced924ee145ec448e3f0d5c8f3a2133ef6ba)). This just a proof-of-concept, but solves my immediate issue.

This provider already has [conditional features depending on the version](https://github.com/cyrilgdn/terraform-provider-postgresql/blob/a961e75c59b360f901020c612deef1e1da8ecfea/postgresql/config.go#L23), so my question is, is it reasonable to expand this to support other "near-Postgres" databases? So Cockroach is one, Yugabyte another.

Options I can picture:
* This is beyond the scope of this Provider, and this will target vanilla Postgres - thus a fork would be reasonable.
* Extend the conditional features to not [just be predicated on the Postgres version](https://github.com/cyrilgdn/terraform-provider-postgresql/blob/a961e75c59b360f901020c612deef1e1da8ecfea/postgresql/config.go#L68), but also on the type of postgres implementation (vanilla vs. cockroach vs. yugabyte vs...).

To add a few notes about how the conditional features could look:
* The provider right now either deduces the version from `SELECT version();`, which does not work for Cockroach at least, but users can set `expectedVersion`. Ideally, parsing for Cockroach and other versions would be supported.
* The provider would also take a parameter of which implementation it was using - i.e. vanilla or cockroachdb or ...
* Thus the 'version' would not be the postgres version, but the version of the implementation. This would mean every feature would say, for each implementation, what version it is available at (if any).

Of course, there are questions about the resilience of these other implementations (e.g. adding integration testing with each one).

Thoughts?

**Comment by NidalShaterM** [comment]
@kylepl, did you publish a TF provider for this, I'm trying to replicate your changes and publish a new provider but I'm facing some issues with publishing it
@cyrilgdn could you tell me if you are planning to support cockroachDB soon?

**Comment by kylepl** [comment]
@NidalShaterM - I did not, though I still use it locally. I never investigated how to publish it, so if it's actually the publishing part, I won't be much assistance, if it's something else, feel free to comment here.

**Comment by NidalShaterM** [comment]
> @NidalShaterM - I did not, though I still use it locally. I never investigated how to publish it, so if it's actually the publishing part, I won't be much assistance, if it's something else, feel free to comment here.

@kylepl, thanks for the quick response, I was trying to modify and use the provider with your changes, so I forked this repo and tried to use github directly as a source in the required_provider
```
terraform {
  required_providers {
    postgresql = {
      source = "github.com/nidalshaterm/postgresql"
      version="1.22.4"
    }
  }
}
terraform init

Initializing the backend...

Initializing provider plugins...
- Finding github.com/nidalshaterm/postgresql versions matching "1.22.4"...
╷
│ Error: Invalid provider registry host
│ 
│ The given source address "github.com/nidalshaterm/postgresql" specifies a GitHub repository rather than a Terraform provider. Refer to the documentation of the provider to find
│ the correct source address to use.
```
I tried also to use git submodules:
```
git submodule add https://github.com/NidalShaterM/terraform-provider-postgresql.git providers/postgressql
git submodule init  
 git submodule update
 ```
 then 
 ```
 terraform {
  required_providers {
    postgresql = {
      source  = "./providers/postgressql"
      version="1.22.4"
    }
  }
}
terraform init                                                                                           

Initializing the backend...
Terraform encountered problems during initialisation, including problems
with the configuration, described below.

The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
╷
│ Error: Invalid provider source hostname
│ 
│   on main.tf line 4, in terraform:
│    4:       source  = "./providers/postgressql"
│ 
│ Invalid provider source hostname namespace "" in source "./providers/postgressql": hostname contains empty label (two consecutive periods)"
╵

╷
│ Error: Invalid provider source hostname
│ 
│   on main.tf line 4, in terraform:
│    4:       source  = "./providers/postgressql"
│ 
│ Invalid provider source hostname namespace "" in source "./providers/postgressql": hostname contains empty label (two consecutive periods)"
╵
```

I tried to publish my provider and had the same error reported here 
https://discuss.hashicorp.com/t/unable-to-publish/18757

so is there a fast way to test the CockroachdB changes?

**Comment by NidalShaterM** [comment]
adding `expected_version = 13` worked, thanks for your help

**Comment by kylepl** [comment]
Agreed that there is likely non-trivial development to make it work well, I hadn't, for instance, tried removing a user.
